sklearn/datasets/value_dropper.py

```diff
@@ -0,0 +1,395 @@
+                elif isinstance(val, (np.ndarray, list, tuple)):
+                    val = np.asarray(val)
+                    if val.shape[0] != n_features:
+                        raise ValueError("The shape of the per feature"
+                                         " drop probabilities vector "
+                                         "for label, %s, does not conform"
+                                         " to the number of features, %d"
+                                         % (item, n_features))
+                    drop_probs[item, :] = val
+                else:
+                    raise ValueError("If missing_distribution is a dict with"
+                                     " target labels as keys, the values of "
+                                     "the dict should either be a single float"
+                                     " or an array of shape (n_features,). "
+                                     "%r was passed for label %r"
+                                     % (val, item))
+
+        else:
+            # For MCAR
+
+            # Validate and convert the missing_distribution dict into a
+            # 1D probability distribution along the features
+
+            drop_probs = np.zeros(n_features,
+                                  dtype=np.float64).reshape((1, n_features))
+
+            if isinstance(self.missing_distribution, (float, np.floating)):
+                drop_probs[:] = self.missing_distribution / n_features
+            elif isinstance(self.missing_distribution,
+                            (list, tuple, np.ndarray)):
+                missing_distribution = np.asarray(self.missing_distribution)
+                if missing_distribution.shape[0] != n_features:
+                    raise ValueError("The shape of the per feature "
+                                     "drop probabilities vector does not "
+                                     "conform to the number of features, %d"
+                                     % n_features)
+                drop_probs[:] = self.missing_distribution
+            else:
+                raise ValueError("missing_distribution must be a float or "
+                                 " 1D vector (list, tuple or np.ndarray) of "
+                                 "shape (n_features,) or dict of 1D vector / "
+                                 "floats. %r was passed"
+                                 % self.missing_distribution)
+
+        if 1 - drop_probs.ravel().sum() <= np.finfo(float).eps:
+            raise ValueError("The sum of all probabilities in the "
+                             "missing_distribution should sum up to less "
+                             "than 1. "
+                             "The sum was found to be %0.8f"
+                             % drop_probs.ravel().sum())
+
+        drop_counts = (drop_probs * n_values).astype(int)
+
+        def _block_drop_missing_values(block_missing_mask, block_id):
```

Comment:
I'm against function in other function (or only if necessary which is not the case here).